Rate Limiting, Max Concurrency, Infinite Crawl & Additional Configurations

cpdata opened the pull request:
# Initial Improvements

## Main Additions
- `maxPagesToCrawl` if = 0 then will crawl all matching urls and display during progress as 1/∞.
- `maxConcurrency` Sets the number of concurrent crawl requests. If left unset then the `undefined` maxConcurrency will do maximum parallel connections like the originals default. Now defaults to 1 to avoid getting IP banned.
- `waitPerPageCrawlTimeoutRange` Defaults to a range of 1 second to 1 second but can be set to create a random delay between any 2 numbers in milliseconds to avoid rate limit rejection when crawling.
- `headless` is `true` by default but can now be configured in the config.ts file for situations that require it.
- Improved README.md & config.ts documentation. ( More to be done.)

## Full Summery 
- Added *.code-workspace to .gitignore for VSCODE workspaces saved in the root of the project.
[Add VSCode workspace file in .gitignore](https://github.com/BuilderIO/gpt-crawler/commit/dbb6a21b025f93791e2675bef707eb07dcb7ddb8)
- Final output .json files go to `outputs/` folder so they are not overwritten.
[Add outputs dir to .gitignore for final outputs](https://github.com/BuilderIO/gpt-crawler/commit/60ec1883756c7a838d95d347888e25dfc0d4b85d)
- Dynamic domain + date-timestamp final output file name  ex. outputs/domain.com-2023-11-28-12:02:51.json
[Add Dynamic OutputFileName based on date-timestamp](https://github.com/BuilderIO/gpt-crawler/commit/14eb9fab1e74d952ee8b48813b2fc00e51865a96)
- `maxPagesToCrawl`: if set to 0 will continue crawling for all matching URLs and display infinity symbol ex. 1/∞, 2/∞, 3/∞ etc.( default = 50 )
[Allow maxPagesToCrawl to be optional and infinite by setting 0 which will display the infinity symbol](https://github.com/BuilderIO/gpt-crawler/commit/e700f6e14db09e225bf594307dc59bcce3873c2a) 
- `maxConcurrency`: Some sites will automatically block connections to prevent DDOS attacks. This config sets how many concurrent requests run at a time. ( default = 1 )
[Added maxConcurrency config to set maximum concurrent parallel requests.](https://github.com/BuilderIO/gpt-crawler/commit/ac0ac25a3240bd57f392f8655d43af2436b93777)
- Updates to core.ts to add config paramters for maxPagesToCraw, maxConcurrency, maxRequestsPerCrawl, headless
- `waitPerPageCrawlTimeoutRange` config added to set a random range in milliseconds between requests. Some sites will automatically block connections so this is a 2 number object that introduces a random delay between requests for rate limit handling( default = 1000 )
[Update to core.ts for maxPagesToCrawl](https://github.com/BuilderIO/gpt-crawler/commit/a6b4b1f61119340b97dc0776d1cc45924e08b354)
- `headless` is now a config option ( default = true )
[Addded headless mode as a configuration parameter](https://github.com/BuilderIO/gpt-crawler/commit/c6b63036a9615550b27eda09debeb746b569b7ee)
- Random Rate Limiting Range with `waitPerPageCrawlTimeoutRange` config.
[Added waitPerPageCrawlTimeoutRange for a random range in milliseconds between page requests to help with rate limiting](https://github.com/BuilderIO/gpt-crawler/commit/b427b253951fb403fff7ccc3c763f6cc8db02134)
- 1 line improvement to prevent VSCODE warning for non-existent docker container.
[Added ts-ignore for docker config.ts to prevent VSCode from declaring missing file that isn't created until the Docker is.](https://github.com/BuilderIO/gpt-crawler/commit/35ea95bdd8f42d137a11b732a8249b4c72e722e3)
- Chunked data goes into the `storage` dir. Final compiled JSON file outputs go into the new `outputs` directory.
[Added Output Directory for all outputFileName to go into so they aren't overwritten in storage](https://github.com/BuilderIO/gpt-crawler/commit/a996ab125b4af4bb332beea6f4159a728b77ce26)
- Added more variables to the ./config.ts file for setting up the config in a more customized way that also includes the automatic naming convention domain-timestamp.json
[Additions to dynamic url and match configurations in config.ts](https://github.com/BuilderIO/gpt-crawler/commit/33faaf5a89223cf2e64493406b39866fc5368ad0)
- Added details for waitForSelectorTimeout in the README.md file
[Added waitForSelectorTimeout to README.md](https://github.com/BuilderIO/gpt-crawler/commit/83a5b0cd42d37df6f2e7f74e43cbf6bc6797bbec)
- Added additional Markdown and Typescript formatting to the config.ts and README.md files.
[Adding details to README.md and config.ts as well as extra formatting.](https://github.com/BuilderIO/gpt-crawler/commit/401fa9bf55599c5bea2c3a6e27bb113223771630)


- 13 Commits hopefully makes review a little easier. 

I would like to contribute to this project on a regular basis. I have a lot of Web-scraping, A.I./LLMs, CI/CD, Automation, experience and would like to discuss with the main collaborators and see were I can be of the most use.

cpdata committed:
Add VSCode workspace file in .gitignore

cpdata committed:
Add outputs dir to .gitignore for final outputs

cpdata committed:
Add Dynamic OutputFileName based on date-timestamp

cpdata committed:
Allow maxPagesToCrawl to be optional and infinite by setting 0 which will display the infinity symbol. Default is 50

cpdata committed:
Added maxConcurrency config to set maximum concurrent parallel requests.

cpdata committed:
Update to core.ts for maxPagesToCrawl

cpdata committed:
Addded headless mode as a configuration parameter

cpdata committed:
Added waitPerPageCrawlTimeoutRange for a random range in milliseconds between page requests to help with rate limiting

cpdata committed:
Added ts-ignore for docker config.ts to prevent VSCode from declaring missing file that isn't created until the Docker is.

cpdata committed:
Added Output Directory for all outputFileName to go into so they aren't overwritten in storage

cpdata committed:
Additions to dynamic url and match configurations in config.ts

cpdata committed:
Added waitForSelectorTimeout to README.md

cpdata committed:
Adding details to README.md and config.ts as well as extra formatting.

marcelovicentegc commented on README.md:
```diff
@@ -64,22 +64,31 @@ export const defaultConfig: Config = {
 type Config = {
-  /** URL to start the crawl, if sitemap is provided then it will be used instead and download all pages in the sitemap */
+  
+  /** Required - URL to start the crawl, if sitemap is provided then it will be used instead and download all pages in the sitemap */
```
What do you think of using tags instead, for example:

```suggestion
  /** 
  * URL to start the crawl, if sitemap is provided then it will be used instead and download all pages in the sitemap 
  * @required
  */
```

This way we could eventually bring jsdoc/typedoc into the mix to generate meaningful documentation from them. Just a suggestion, as even jsdoc/typedoc can infer whether a property is required or not based on its typings 🤗

marcelovicentegc commented on config.ts:
```diff
@@ -1,8 +1,32 @@
+const __dirname = dirname(__filename);
+
+const starting_url = "https://www.builder.io/c/docs/developers";
+const url_prefix = "https://"
+const domain = "www.builder.io";
+const url_suffix = "/c/docs";
+const base_url = url_prefix + domain;
+const match_url_prefix = base_url + url_suffix;
+const match_url = match_url_prefix + "/**";
+
+// Now date stamp for output file name
+const now = new Date();
+const date = now.toISOString().split('T')[0];
+const time = now.toTimeString().split(' ')[0];
+const outputs_dir = __dirname.split('/').slice(0, -1).join('/') + '/outputs';
```
Just a nitpick here to guarantee code consistency, let's use camelCase for these variables:

```suggestion
const urlPrefix = "https://"
const domain = "www.builder.io";
const urlSuffix = "/c/docs";
const baseUrl = urlPrefix + domain;
const matchUrlPrefix = baseUrl + urlSuffix;
const matchUrl = matchUrlPrefix + "/**";

// Now date stamp for output file name
const now = new Date();
const date = now.toISOString().split('T')[0];
const time = now.toTimeString().split(' ')[0];
const outputsDir = __dirname.split('/').slice(0, -1).join('/') + '/outputs';
```

> Requires changes below 👇

cpdata replied:
I don't mind nitpicks 👍establishing a shared nomenclature for project is ideal in my opinion. I am always open to feedback and recommendations.

marcelovicentegc commented on src/config.ts:
```diff
@@ -6,43 +6,54 @@ const Page: z.ZodType<Page> = z.any();
 
 export const configSchema = z.object({
   /**
+   * **Required:**
```
Same from my comment on README 😉

cpdata replied:
I used the same convention as mentioned for the README.md file for the src/config.ts file as well. I will add more details for parameters, examples, defaults, etc. as we progress.

cpdata committed:
Update prettier formatting for README.md, src/config.ts, src/core.ts, and config.ts and formatting for jsdoc/typedoc as  recommened by @marcelovicentegc in pull request #102, added .prettierignore file

cpdata commented:
I updated with prettier formatting for the files that failed README.md, src/config.ts, src/core.ts, and config.ts.
I also added the formatting for jsdoc/typedoc as  recommened by @marcelovicentegc in response to my orginal pull request #102. Additionally, I added .prettierignore file.

steve8708 commented:
@marcelovicentegc this look good to you to merge?

marcelovicentegc commented on config.ts:
```diff
@@ -1,8 +1,33 @@
+const domain = "www.builder.io";
+const urlSuffix = "/c/docs";
+const baseUrl = urlPrefix + domain;
+const matchUrl_prefix = baseUrl + urlSuffix;
+const matchUrl = matchUrl_prefix + "/**";
+
+// Now date stamp for output file name
+const now = new Date();
+const date = now.toISOString().split("T")[0];
+const time = now.toTimeString().split(" ")[0];
+const outputs_dir = __dirname.split("/").slice(0, -1).join("/") + "/outputs";
+
+const outputFileName =
+  outputs_dir + "/" + domain + "-" + date + "-" + time + ".json";
```
Hey @cpdata! Just a couple more nitpicks (replacing snake_case for camelCase) and a rebase and we are good to go! 
```suggestion
const matchUrlPrefix = baseUrl + urlSuffix;
const matchUrl = matchUrlPrefix + "/**";

// Now date stamp for output file name
const now = new Date();
const date = now.toISOString().split("T")[0];
const time = now.toTimeString().split(" ")[0];
const outputsDir = __dirname.split("/").slice(0, -1).join("/") + "/outputs";

const outputFileName =
  outputsDir + "/" + domain + "-" + date + "-" + time + ".json";
```

marcelovicentegc commented:
> @marcelovicentegc this look good to you to merge?

Hey @steve8708! Happy new years! One rebase and a few nitpicks ☝️ and it occurs to me that we are good to go 🤗

Ademrobert commented:
Please merge this branch ASAP!

zanamich commented:
@cpdata Any plans to merger this soon?

steve8708 commented:
Looks like @marcelovicentegc isn’t responding so I’m happy to merge this once the branch is in a mergeable state - apologies for the delay all

cpdata commented:
@zanamich @steve8708 Thanks guys for the heads up I didn't realize this was never merged. I'll look into it this weekend.

zanamich commented:
@cpdata Any luck? :)
